File: Concepts/007.recursion/01GenerateParenthesis/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();

        generate(0, 0, n, "", result);
        return result;
    }

    private void generate(int open, int close, int n, String current, List<String> result) {
   
        if (open == close && open + close == 2 * n) {
            result.add(current);
            return;
        }

        if (open < n) {
            generate(open + 1, close, n, current + '(', result);
        }

    
        if (close < open) {
            generate(open, close + 1, n, current + ')', result);
        }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int n = 3; // Example input
        List<String> result = sol.generateParenthesis(n);

        System.out.println("All combinations of balanced parentheses for n = " + n + " are:");
        for (String combination : result) {
            System.out.println(combination);
        }
    }
}

/*
Output:
All combinations of balanced parentheses for n = 3 are:
((()))
(()())
(())()
()(())
()()()
*/

File: Concepts/007.recursion/01GenerateParenthesis/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
private: 
    void func(int open, int close, int n, string s, vector<string> &ans) {

        if (open == close && (open + close) == 2 * n) {
            ans.push_back(s); 
            return; 
        }
        
        if (open < n) {
            func(open + 1, close, n, s + '(', ans); 
        }
        

        if (close < open) {
            func(open, close + 1, n, s + ')', ans); 
        }
    }
public:

    vector<string> generateParenthesis(int n) {

        vector<string> ans; 
     
        func(0, 0, n, "", ans); 
        return ans; 
    }
};

int main() {
    Solution sol;
    int n = 3; // Example input
    vector<string> result = sol.generateParenthesis(n);

    cout << "All combinations of balanced parentheses for n = " << n << " are:" << endl;
    for (const string &combination : result) {
        cout << combination << endl;
    }

    return 0;
}

/*
All combinations of balanced parentheses for n = 3 are:
((()))
(()())
(())()
()(())
()()()
*/